Add TPU support for basic_training

ianstenbit opened the pull request:
@LukeWood

martin-gorner commented on examples/training/classification/imagenet/basic_training.py:
```diff
@@ -142,9 +145,16 @@ def augment(img, label):
-# For TPU training, use tf.distribute.TPUStrategy()
-# MirroredStrategy is best for a single machine with multiple GPUs
-strategy = tf.distribute.MirroredStrategy()
+
```
2 comments:

1) you are using old code for instantiating a TPUStrategy. There is a shorter version now: TPUClusterResolver.connect(). See example here: https://bit.ly/keras-TPU

2) it is always more user-friendly to have the code detect the hardware automatically (instead of a flag). This is easy to do with distribution strategies. Also see: https://bit.ly/keras-TPU

martin-gorner reviewed (changes requested):
With Distribution Strategies the hardware can be detected automatically.

ianstenbit commented:
/gcbrun

LukeWood commented:
> With Distribution Strategies the hardware can be detected automatically.

if hardware can (optimally) be detected, should tf.distribute offer this API?

@rchao @qlzh727 may have some insight in here

rchao commented:
I know there's been plan for auto distribution. Cc'ing @yuefengz for the latest progress.

martin-gorner commented:
1) when changing the code, you also deleted your helpful comment. I would leave it in:
"MirroredStrategy is best for a single machine with one or multiple GPUs."

2) Did you test this on TPU, multi-GPU ? Does it work ? does it actually accelerate the training ?

3) If you want the additional hardware to actually accelerate something (Note: a single TPU has 8 cores), you have to scale the batch size and the learning rate with the number of replicas. Something like:
BATCH_SIZE = base_batch_size * strategy.num_replicas_in_sync
LR = base_lr * strategy.num_replicas_in_sync
But it's not a hard rule, only a good starting point to find the best batch size and learning rate (if you are working with an LR schedule, you have to scale the entire schedule). 
The theory is that with more hardware, you increase the batch size to process more images at once. However you are still doing only one gradient update per batch. So if you want training to actually progress faster, each gradient update must be moving the weights more, that's why you also increase the LR.

martin-gorner commented:
4) for data loading, would you consider slightly changing the code to make it transparent for users and help those who want to plug their own dataset ? I put some recommendations here:
https://docs.google.com/document/d/1o1cqRgPhUYfYwgWGBPkAB0nble5TLpYXfH0sQx3c25A/edit?resourcekey=0-zqwbubVFUQYiSfwxYKhqhA#bookmark=id.nrepz1rgy2zm
In short, after the tfds.load call, which prepares the dataset as TFRecords, load from the TFRecords using explicit code. It's only a couple of lines of additional code but it makes the code easy to repurpose for your own custom dataset.

ianstenbit commented:
I've re-added the strategy comments, and yes I've verified that this works on TPU and multi-GPU systems (both on GCP)

3. These are both flag-controlled, and I think it may get confusing if we modify what is passed in to the flag. In particular, this gets confusing for our scores metadata, as the metadata contains the values of batch_size and LR that were passed using the flag, but if we mutate them they may no longer be correct in the metadata. I think letting the user specify this via CLI arguments to the script resolves this issue and still allows flexibility to adapt batch size and LR to number of accelerators.

4. I am happy to update this in a separate PR. I also think this might be a worthy topic for a Keras or KerasCV tutorial.

martin-gorner commented:
Scaling the batch size and learning rate schedule appropriately for a given
cluster size is the hard part of scaling. I don't think we should be
leaving that to users. Most users will assume that this is supposed to just
work and fail to get a speedup with more hardware.

On Wed, 7 Sept 2022 at 18:14, Ian Stenbit ***@***.***> wrote:

> I've re-added the strategy comments, and yes I've verified that this works
> on TPU and multi-GPU systems (both on GCP)
>
>    1.
>
>    These are both flag-controlled, and I think it may get confusing if we
>    modify what is passed in to the flag. In particular, this gets confusing
>    for our scores metadata, as the metadata contains the values of batch_size
>    and LR that were passed using the flag, but if we mutate them they may no
>    longer be correct in the metadata. I think letting the user specify this
>    via CLI arguments to the script resolves this issue and still allows
>    flexibility to adapt batch size and LR to number of accelerators.
>    2.
>
>    I am happy to update this in a separate PR. I also think this might be
>    a worthy topic for a Keras or KerasCV tutorial.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keras-team/keras-cv/pull/757#issuecomment-1239604112>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHKKZ2YHBTZ2I37H3ZW4UDV5C5O7ANCNFSM6AAAAAAQCYGBKQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 

Martin Görner | Product Manager, TensorFlow & Keras | ***@***.*** | +1
425 273 0605

bhack commented:
> Scaling the batch size and learning rate schedule appropriately for a given
cluster size is the hard part of scaling. I don't think we should be
leaving that to users. Most users will assume that this is supposed to just
work and fail to get a speedup with more hardware.

Similar to  https://github.com/Lightning-AI/lightning/issues/11012 but still not supported in DDP (Distributed)

ianstenbit commented:
> Scaling the batch size and learning rate schedule appropriately for a given cluster size is the hard part of scaling. I don't think we should be leaving that to users. Most users will assume that this is supposed to just work and fail to get a speedup with more hardware.

I agree with this for code that's shipped in the package -- but for code in `examples/` that is for training our models, I'm less sure. Having said that, I think you have more context here so I've updated the code to use this approach.

I've also updated the help strings and defaults for the affected flags to reflect the change.

ianstenbit commented:
I've now verified that LR and batch size scaling is working on a system with 2 GPUs and on a TPU with 8 cores.

martin-gorner commented:
great!

ianstenbit commented:
/gcbrun
